packages/arb-validator/rollupvalidator/txtracker.go

```diff
@@ -110,214 +136,290 @@ func (a *assertionInfo) FindLogs(address *big.Int, topics []common.Hash) []logRe
 				}
 			}
 			if match {
-				logs = append(logs, logResponse{evmLog, txLogs.msg})
+				logs = append(logs, logResponse{
+					Log:     evmLog,
+					TxIndex: txLogs.TxIndex,
+					TxHash:  txLogs.TxHash,
+				})
 			}
 		}
 	}
 	return logs
 }
 
-func newAssertionInfo() *assertionInfo {
-	return &assertionInfo{}
-}
-
 type txTracker struct {
-	txRequestIndex int
-	transactions   map[common.Hash]txInfo
-	assertionInfo  []*assertionInfo
-	assertionMap   map[common.Hash]*assertionInfo
-	accountNonces  map[common.Address]uint64
-	vmID           common.Address
-	requests       chan validatorRequest
+	sync.RWMutex
+	rollup.NoopListener
+	txDB          *txDB
+	maxNodeHeight uint64
+	chainAddress  common.Address
 }
 
 func newTxTracker(
-	vmID common.Address,
-) *txTracker {
-	requests := make(chan validatorRequest, 100)
+	db machine.CheckpointStorage,
+	ns machine.NodeStore,
+	chainAddress common.Address,
+) (*txTracker, error) {
+	txdb, err := newTxDB(db, ns, chainAddress)
+	if err != nil {
+		return nil, err
+	}
 	return &txTracker{
-		txRequestIndex: 0,
-		transactions:   make(map[common.Hash]txInfo),
-		assertionInfo:  make([]*assertionInfo, 0),
-		assertionMap:   make(map[common.Hash]*assertionInfo),
-		accountNonces:  make(map[common.Address]uint64),
-		vmID:           vmID,
-		requests:       requests,
+		txDB:          txdb,
+		chainAddress:  chainAddress,
+		maxNodeHeight: 0,
+	}, nil
+}
+
+// Delete assertion and transaction data from the reorged blocks if there are any
+func (tr *txTracker) RestartingFromLatestValid(_ context.Context, chain *rollup.ChainObserver, node *structures.Node) {
+	startDepth := node.Depth()
+	go func() {
+		tr.Lock()
+		defer tr.Unlock()
+		// First remove any data from reorged nodes
+		for i := tr.maxNodeHeight; i > startDepth; i-- {
+			if err := tr.txDB.removeUnconfirmedNode(i); err != nil {
+				continue
+			}
+		}
+
+		tr.maxNodeHeight = startDepth
+
+		// Next process data for any nodes which have not yet been processed
+		chain.ReplayNodesToLatestValid(tr.processNextNode)
+	}()
+}
+
+func (tr *txTracker) AdvancedKnownNode(_ context.Context, _ *rollup.ChainObserver, node *structures.Node) {
+	go tr.processNextNode(node)
+}
+
+func (tr *txTracker) ConfirmedNode(_ context.Context, _ *rollup.ChainObserver, ev arbbridge.ConfirmedEvent) {
+	go func() {
+		tr.Lock()
+		defer tr.Unlock()
+
+		if err := tr.txDB.confirmNode(ev.NodeHash); err != nil {
+			log.Println(err)
+			return
+		}
+	}()
+}
+
+func (tr *txTracker) processNextNode(node *structures.Node) {
```

Comment:
Is there some locking needed here?  We could have multiple instances to processNextNode running at the same time, and other stuff happening simultaneously as well.

Also concerned about race conditions between this async thread and reorgs.

Reply:
There's a write lock that gets held by this function after it finishes processing the node which is a pure function so I think this is safe as is

Reply:
My bad.  There was locking there.